proc-macros/src/lib.rs

```diff
@@ -89,37 +57,22 @@ pub fn rpc_api(input_token_stream: TokenStream) -> TokenStream {
 }
 
 /// Generates the macro output token stream corresponding to a single API.
-fn build_api(api: api_def::ApiDefinition) -> Result<proc_macro2::TokenStream, syn::Error> {
+fn build_client_api(api: api_def::ApiDefinition) -> Result<proc_macro2::TokenStream, syn::Error> {
 	let enum_name = &api.name;
-
-	// TODO: make sure there's no conflict here
-	let mut tweaked_generics = api.generics.clone();
-	tweaked_generics
-		.params
-		.insert(0, From::from(syn::LifetimeDef::new(syn::parse_str::<syn::Lifetime>("'a").unwrap())));
-	tweaked_generics.params.push(From::from(syn::TypeParam::from(syn::parse_str::<syn::Ident>("R").unwrap())));
-	tweaked_generics.params.push(From::from(syn::TypeParam::from(syn::parse_str::<syn::Ident>("I").unwrap())));
-	let (impl_generics, ty_generics, where_clause) = tweaked_generics.split_for_impl();
-	let generics = api
-		.generics
-		.params
-		.iter()
-		.filter_map(|gp| if let syn::GenericParam::Type(tp) = gp { Some(tp.ident.clone()) } else { None })
-		.collect::<HashSet<_>>();
-
 	let visibility = &api.visibility;
+	// TODO: make sure there's no conflict here
+	let tweaked_generics = api.generics.clone();
+	let mut non_used_type_params = HashSet::new();
 
 	let mut variants = Vec::new();
-	let mut tmp_variants = Vec::new();
 	for function in &api.definitions {
-		let function_is_notification = function.is_void_ret_type();
 		let variant_name = snake_case_to_camel_case(&function.signature.ident);
-		let ret = match &function.signature.output {
-			syn::ReturnType::Default => quote! {()},
-			syn::ReturnType::Type(_, ty) => quote_spanned!(ty.span()=> #ty),
+		if let syn::ReturnType::Type(_, ty) = &function.signature.output {
+			non_used_type_params.insert(ty);
```

Comment:
return types are not bound, so handle them differently in a separate enum variant

such as: `Foo { _0: T, _1: U, .... }` I'm no proc macro expert but I realized that this may make the enum size bigger if this variant grows huge :)